### DISCUSSION TITLE
Integration for phpIPAM

### DISCUSSION THREAD

**Discussion by steled** [kubermatic/kubeone#2603] [Q&A]
Hi,

we are trying to use KubeOne to set up our cluster via vSphere.
Unfortunately we are not using DHCP server, we are only working with phpIPAM.
All our tools for setting up VMs rely on an integration with phpIPAM.

So my quesiton, how can we use KubeOne to spin up the VMs by using phpIPAM.
Currently I didn't found an integration for this or an other documentation or hints.

Maybe someone can give a tip how we can solve this.

Thanks & best regards

**Comment by embik** [comment]
Hey @steled,

I'm not familiar with phpIPAM and how integration for VMs looks in practice, but have you taken a look at [Operating System Manager (OSM)](https://docs.kubermatic.com/kubeone/v1.5/architecture/operating-system-manager/)'s feature for providing custom user-data to VMs via `OperatingSystemProfiles`? There's more information [here](https://docs.kubermatic.com/kubeone/v1.5/architecture/operating-system-manager/usage/#using-custom-operatingsystemprofile) for the usage in KubeOne and [here](https://docs.kubermatic.com/operatingsystemmanager/guides/customize-operating-system-profile/).

This would mean that your VMs would need to self-register with phpIPAM when booting with some commands included in the custom profile.

If that doesn't work, the best way to go is probably modifying the Terraform files to run your necessary steps while creating the VMs. That would mean you need to configure static workers and machine-controller would likely not be usable.

**Reply by moadqassem** [reply]
As what @embik already suggested, you can create a custom OSP that can talk to the phpIPAM(it has a REST-API AFAIK), acquire all the network configs, ip, gateway, netmask, etc...and then use these configs to provision the networking part for the VM. I would suggest that create bash script, to get these info and save it as a netplan file, and then execute netplan in another script if needed.

**Reply by steled** [reply]
Thanks for your reply and hints.

But this means that the VM that is configured by ths OSM needs access to the phpIPAM right?
The problem is that after starting the node VMs they don't have an IP assigned to.
First we need to get the IP from phpIPAM and than assign it to the node VM.

**Reply by steled** [reply]
Is there a genereal IPAM integration available?

**Reply by moadqassem** [reply]
Damn, you are actually correct. I overlooked that chicken and egg problem 😅, However there is a super limited solution in machine controller. You can actually assign a static IP range in a machine deployment using something like this: https://github.com/kubermatic/machine-controller/blob/main/test/e2e/provisioning/testdata/machinedeployment-vsphere-static-ip.yaml#L45-L51 

However, this is only supported by flatcar and I am not sure how much it is compatible with new ignition provisioners. 

P.S : AFAIK, No we don't have support for generic IPAMs

**Reply by steled** [reply]
ok thanks for your answer.... 

bad for us, so we have to see how we can solve this problem or have to use static workers :see_no_evil:

**Reply by moadqassem** [reply]
Sorry about that 😞 however PRs are welcome 😉
